src/infra/core/src/utils/datasets_filtering.rs

```diff
@@ -48,3 +51,48 @@ pub fn filter_datasets_by_pattern(
             .boxed()
     }
 }
+
+pub fn filter_datasets_by_any_pattern(
+    dataset_repo: &dyn DatasetRepository,
+    dataset_ref_patterns: Vec<DatasetRefPatternAny>,
+) -> StreamMap<&'_ str, FilteredDatasetRefAnyStream<'_>> {
+    let mut is_pattern = false;
+    let mut static_dataset_refs = vec![];
+    for dataset_ref_pattern in &dataset_ref_patterns {
+        if dataset_ref_pattern.is_pattern() {
+            is_pattern = true;
+        } else {
+            static_dataset_refs.push(Ok(dataset_ref_pattern
+                .as_dataset_ref_any()
+                .unwrap()
+                .clone()));
+        }
+    }
+
+    let mut result = StreamMap::new();
+
+    result.insert(
+        "static_dataset_refs",
+        Box::pin(stream::iter(static_dataset_refs)) as FilteredDatasetRefAnyStream,
+    );
+
+    if is_pattern {
+        result.insert(
+            "filtered_dataset_refs",
+            dataset_repo
+                .get_all_datasets()
+                .try_filter(move |dataset_handle| {
+                    future::ready(
+                        dataset_ref_patterns.iter().any(|dataset_ref_pattern| {
+                            dataset_ref_pattern.is_match(dataset_handle)
+                        }),
+                    )
+                })
+                .map_ok(|dataset_handle| dataset_handle.as_any_ref())
+                .map_err(Into::into)
+                .boxed(),
+        );
+    }
+
+    result
+}
```

Comment:
I would suggest to simplify return type to be a simple stream.

Internally you coud simply concat 2 streams, I don't understand any need in this static string marker.

Stream 1:
```
Box::pin(async_stream::try_stream! {
   //Yield all non-pattern any-refs.
}
```

Stream2: If any patterns present, at stage 2, iterate datasets and call `is_match`.

Concatenate stream 1 and stream 2 with [`chain`](https://dtantsur.github.io/rust-openstack/tokio/stream/trait.StreamExt.html#method.chain)